subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/validation/MavenPublicationErrorChecker.java

```diff
@@ -43,14 +48,56 @@ public abstract class MavenPublicationErrorChecker extends PublicationErrorCheck
      * @throws PublishException if the artifacts are modified
      */
     public static void checkThatArtifactIsPublishedUnmodified(PublishArtifact source, DefaultMavenArtifactSet mainArtifacts) {
+        // Do we want to allow _additional_ artifacts to published? If not, we need to completely associate the source and artifacts
+        // at a higher level. As-is, this just verifies that none are removed from the components artifacts.
+        MavenArtifact bestMatch = null;
+        Set<Difference> bestMatchDifferences = null;
         for (MavenArtifact mavenArtifact : mainArtifacts) {
-            if (source.getFile().equals(mavenArtifact.getFile())
-                && source.getExtension().equals(mavenArtifact.getExtension())
-                && Strings.nullToEmpty(source.getClassifier()).equals(Strings.nullToEmpty(mavenArtifact.getClassifier()))) {
+            EnumSet<Difference> difference = EnumSet.noneOf(Difference.class);
+            if (!source.getFile().equals(mavenArtifact.getFile())) {
+                difference.add(Difference.FILE);
+            }
+            if (!source.getExtension().equals(mavenArtifact.getExtension())) {
+                difference.add(Difference.EXTENSION);
+            }
+            if (!Strings.nullToEmpty(source.getClassifier()).equals(Strings.nullToEmpty(mavenArtifact.getClassifier()))) {
+                difference.add(Difference.CLASSIFIER);
+            }
+            // If it's all equal, we found a matching artifact that is being published
+            if (difference.isEmpty()) {
                 return;
             }
+            if (bestMatch == null || difference.size() < bestMatchDifferences.size()) {
+                bestMatch = mavenArtifact;
+                bestMatchDifferences = difference;
```

Comment:
🤔 Is there a better way to determine this?  The extension check will always fail along with file check - so the same file name published under a different classifier will always win.  With multiple artifacts with similar names, we'd probably want to do some sort of String difference on the names.  I'm not sure this is a reliable way to determine the best match.

The list of available artifacts should be small - I'd consider just emitting the entire list of filenames here and letting the build author figure out which one is the best.

Reply:
>  The extension check will always fail along with file check - so the same file name published under a different classifier will always win.

This is not necessarily true, see my other comment / tests.

It might indeed just be a better idea to print all file names, but I worry about emitting too much and overwhelming the user. Isn't the expected fix here to simply _stop touching artifacts at all_, in which case you don't particularly care which one it is too much?

Reply:
> Isn't the expected fix here to simply stop touching artifacts at all, in which case you don't particularly care which one it is too much?

Yes, assuming you can do this.

This could do something like print the first 3 artifacts, then if there are more, display  a "... (only the first 3 artifacts are shown)".  My concern is that this picks the wrong "best match" and really confuses an author wondering how to align that file with their expectation.